aten/src/ATen/native/BlasKernel.cpp

```diff
@@ -356,4 +358,84 @@ INSTANTIATE_VDOT_IMPL(c10::complex<double>);
 
 #undef INSTANTIATE_DOT_IMPL
 
+namespace blas_impl {
+
+template <typename scalar_t>
+void addr_naive(int64_t m, int64_t n, scalar_t alpha, scalar_t *x, int64_t incx, scalar_t *y, int64_t incy, scalar_t *a, int64_t lda) {
```

Comment:
Ouch. Naive indeed.

Do we have a similar naive ger implementation or does it all go to here?

Follow-up question, how does CUDA handle this?

Reply:
There is no naive `ger` implementation as it is implemented using `addr`.

The CUDA implementation is within the `CUDABlas.cpp` file.
E.g https://github.com/pytorch/pytorch/blob/95a69a7d09e4089dad92e65f5d48ceeec70ba3b4/aten/src/ATen/cuda/CUDABlas.cpp#L374-L382

I believe they are implemented via [cuBLAS](https://docs.nvidia.com/cuda/cublas/index.html) functions, such as the `cublasDgemv`.